exchanges/stream/websocket.go

```diff
@@ -576,183 +555,116 @@ func (w *Websocket) trafficMonitor() {
 	w.Wg.Add(1)
 
 	go func() {
-		var trafficTimer = time.NewTimer(w.trafficTimeout)
-		pause := make(chan struct{})
+		t := time.NewTimer(w.trafficTimeout)
 		for {
 			select {
 			case <-w.ShutdownC:
 				if w.verbose {
-					log.Debugf(log.WebsocketMgr,
-						"%v websocket: trafficMonitor shutdown message received\n",
-						w.exchangeName)
+					log.Debugf(log.WebsocketMgr, "%v websocket: trafficMonitor shutdown message received", w.exchangeName)
 				}
-				trafficTimer.Stop()
+				t.Stop()
 				w.setTrafficMonitorRunning(false)
 				w.Wg.Done()
 				return
-			case <-w.TrafficAlert:
-				if !trafficTimer.Stop() {
-					select {
-					case <-trafficTimer.C:
-					default:
+			case <-time.After(trafficCheckInterval):
+				select {
+				case <-w.TrafficAlert:
+					w.setState(connected)
```

Comment:
I think the only foil to this is how you've changed `w.TrafficAlert` to be a buffered channel. In that context, you actually don't know if the state is connected when you receive a message here, just that at the time the traffic alert was sent, you were connected.

In `connectionMonitor`, when a disconnection error message is received, you set the state to `disconnected`, but no further action is taken until the next cycle until `timer.C` does a check of the connection status and then it will reconnect.

From that, it sounds like there is the possibility of the `TrafficAlert` buffered channel, setting the state to `connected` while disconnected and things attempt to carry on.

When you call `ReadMessage` on a disconnected websocket, my testing says you don't continue to receive `IsDisconnectionError` errors, instead you get the following, so you don't get back to the state "disconnected" until `websocketTrafficTimeout` which default is 30 seconds. So it will eventually reconnect
<img width="810" alt="image" src="https://github.com/thrasher-corp/gocryptotrader/assets/9261323/fd5c289f-9465-4d75-9e5b-fac5af71a38f">

I feel like I'm thinking about this too much though. I like your buffered channel change. I'm not asking you to change this. What are your thoughts on what I'm saying?

Reply:
I think what you're saying makes sense.

Before: Traffic alerts would be sent only when listening, with the impact that we'd miss traffic that might turn out to be important.
Now: Traffic alerts aren't atomic, so we don't know if the sequencing is valid.

Also, I've just realised:
When the traffic Timeout Timer goes off, we might already have traffic, but we're not checking for it.

I'm going to think about this. The tests I have should allow me to iterate a solution.

I already had a thought about solving this problem differently anyway, so I'll take another swing at it.
It means firing a collector goro, which is fine.

I'll get back to you shortly.

Reply:
This hypothesis should be testable.

Split the existing test into behaviours with d6918391 and I'll add a new one now for this behaviour described, which should fail.

Reply:
Fixed 74437455

By never setting the state to connected we avoid any of these issues, and really there's no reason to.

I've reduced the trafficCheckInterval to 100ms.

We also check to make sure any pending traffic signals are seen when we get the trafficTimeout, in case it's out of sequence with trafficCheckInterval

Reply:
Now that its not impacting the connected status, I can understand re-increasing the CheckInterval again. I assume you changed it to this for performance reasons (since we used to act on every single message received)